File: develop/lab-7-end-to-end-app/lab7JavaLab/src/main/java/com/amazonaws/pollynotes/solution/ListFunctionSolution.java
package com.amazonaws.pollynotes.solution;

import java.util.ArrayList;
import java.util.List;

import com.amazonaws.pollynotes.pojo.Note;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBQueryExpression;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;

public class ListFunctionSolution implements RequestHandler<Note, List<Note>> {

	/**
	 * PollyNotes-ListFunction
	 * 
     * This lambda function is integrated with the following API methods:
     * /notes/GET
     *   
     * This function does the following:
     * 1. Takes a JSON payload from API gateway and converts it into a Note POJO {@link com.amazonaws.pollynotes.pojo.Note}
     * 2. Queries DynamoDB based on that payload using the QueryAsync feature
     * 3. Returns the List of Note found
     *  
     * @param	note 	POJO of Note created from the JSON sent from API GW into Lambda
     * @param	context	Lambda context which can be used for logging
     * @return			A List of Note containing all of the elements for the userId in DynamoDB
	 */
	@Override
	public List<Note> handleRequest(Note note, Context context) {
		// The note object only contains the userId as it's the only parameter sent in the /notes/GET
        // {
        //   "userId": "..."
        // }

		context.getLogger().log("Initiating PollyNotes-ListFunction...");
		context.getLogger().log("Note received: {" + note.toString() + "}");
		
		// Create the DynamoDB client and mapper
		AmazonDynamoDB client = AmazonDynamoDBClientBuilder.standard().build();
		DynamoDBMapper mapper = new DynamoDBMapper(client);
		
		// Create the Query Expression with the Hash/Partition Key from the note in the request
		DynamoDBQueryExpression<Note> queryExpression = new DynamoDBQueryExpression<Note>().withHashKeyValues(note);
		
		try {
			// Query DynamoDB based on the Query Expression and return the results
			List<Note> noteList = mapper.query(Note.class, queryExpression);
			
			// Log the Note(s) found
			for (Note n : noteList) {
				context.getLogger().log("Note found in DynamoDB: {" + n.toString() + "}");
			}
			
			// Return the results
			return noteList;
		} catch (Exception e) {
			// Error found, log the message
			context.getLogger().log("Exception found, returning empty list. Error: " + e.getMessage());
			
			// Return an empty list
			return new ArrayList<Note>();
		}
	}
}

File: develop/lab-7-end-to-end-app/lab7CSharpLab/ListFunctionSolution/ListFunctionSolution.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DataModel;
using Amazon.Lambda.Core;
using System.Threading;

// Assembly attribute to enable the Lambda function's JSON input to be converted into a .NET class.
[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.Json.JsonSerializer))]

namespace ListFunctionSolution
{
    public class ListFunctionSolution
    {
        /// <summary>
        /// PollyNotes-ListFunction
        /// 
        /// This lambda function is integrated with the following API methods:
        /// /notes/GET
        /// 
        /// This function does the following:
        /// 
        /// 1. Takes a JSON payload from API gateway and converts it into a Note POCO <see cref="Note"/>
        /// 2. Queries DynamoDB based on that payload using the QueryAsync feature
        /// 3. Returns the List of Note found
        /// </summary>
        /// <param name="note">POCO of Note from the JSON payload from API GW to Lambda. In this case, it will only have the userId set</param>
        /// <param name="context">Lambda context</param>
        /// <returns>List of Note from the Query to DynamoDB based on the userId</returns>
        public List<Note> FunctionHandler(Note note, ILambdaContext context)
        {
            // The note object only contains the userId as it's the only parameter sent in the /notes/GET
            // {
            //   "userId": "..."
            // }

            Console.WriteLine("Initiating PollyNotes-ListFunction...");
            Console.WriteLine("Note received: " + JsonConvert.SerializeObject(note));

            // Create the DynamoDB client and the Context for Object Persistence Model
            AmazonDynamoDBClient client = new AmazonDynamoDBClient();
            DynamoDBContext ddbcontext = new DynamoDBContext(client);

            // Use the QueryAsync method to issue a query to DynamoDB based on the userId
            //  We are using an Async command due to the .netcore SDK which doesn't support sync invokes
            //  We are using the Note POCO defined with the Object Persistence Model so all the data
            //   will be mapped to the right entries. The Table name is also defined in the POCO
            AsyncSearch<Note> noteQuery = ddbcontext.QueryAsync<Note>(note.userId);

            // Retrieve all of the notes based on the Query from DynamoDB and wait
            Task<List<Note>> listTask = noteQuery.GetRemainingAsync();
            listTask.Wait();

            // Once the list has been fetched entirely, check for Exceptions.
            // If there are no Exceptions
            if (listTask.Exception == null)
            {
                Console.WriteLine("Successfully executed QueryAsync with userId: " + note.userId);

                // Log the Notes found
                Console.WriteLine("Notes found in DynamoDB: " + JsonConvert.SerializeObject(listTask.Result));

                // Return the list
                return listTask.Result;
            }
            else
            {
                // There was an exception, log the entry data and the exception
                Console.WriteLine("Unable to QueryAsync note with userId: " + note.userId);
                Console.WriteLine(listTask.Exception);

                // Return an empty list
                return new List<Note>();
            }
        }

        /// <summary>
        /// TESTING ONLY - Run the project as a console application to test the <see cref="FunctionHandler"/>
        /// You should see a list of Note
        /// </summary>
        public static void Main()
        {
            string listTest = @"
            {  
                ""userId"":""testuser""
            }";
            List<Note> noteList = new ListFunctionSolution().FunctionHandler(JsonConvert.DeserializeObject<Note>(listTest), null);
            Console.WriteLine(JsonConvert.SerializeObject(noteList));
        }
    }
}
